### ISSUE TITLE
Generated string tests wrongfully rely on order of maps in go

### ISSUE THREAD

**Issue by romeovs** [gogo/protobuf#342]
I'm working on a project that uses `map<string, string>` field types. For example:

```
option (gogoproto.testgen_all) = true;
option (gogoproto.stringer_all) = true;

message Foo {
  map<string, string> attributes = 1;
}
```

This will generate a type `Foo`:
```
type Foo {
  Attributes map[string]string `protobuf:"..."`
}
```

And a test of the form:
```
func TestFooStringer(t *testing.T) {
	popr := math_rand.New(math_rand.NewSource(time.Now().UnixNano()))
	p := NewPopulatedFoo(popr, false)
	s1 := p.String()
	s2 := fmt.Sprintf("%v", p)
	if s1 != s2 {
		t.Fatalf("String want %v got %v", s1, s2)
	}
}
```

This test is wrong, because it depends on both `s1` and `s2` printing the values in the map in the same order, which is not guaranteed ([source](https://blog.golang.org/go-maps-in-action)):

> When iterating over a map with a range loop, the iteration order is not specified and is not guaranteed to be the same from one iteration to the next.

So the test fails and, even worse, it fails randomly because sometimes the order of the map *is* the same.

The weird thing is that I encountered this issue only for one of the instances in our api where we use `map<string, sting>`, and I've yet to pinpoint why that is. The project is closed source so I'm afraid I can't share the code just yet, and I can't reproduce with a minimal example (although reasoning about the test clearly shows that it should be broken).

**Comment by awalterschulze** [comment]
Ranging over maps is not deterministic in Go.
So the map needs to be sorted before being printed, but from the plugin code, it looks like they are sorted.
https://github.com/gogo/protobuf/blob/master/plugin/stringer/stringer.go#L175

Are you using the newest version of gogoprotobuf?

Hmmm.  This is quite weird.

**Comment by romeovs** [comment]
Yes the keys are sorted in the `Stringer` implementation, but nothing says that the `%v` representation has sorted keys right?

I think the go runtime might do a best effort here, but it can not handle all cases.


I've pinpointed the issue a little more. It only occurs in this case:

```
option (gogoproto.testgen_all) = true;
option (gogoproto.stringer_all) = true;

message Foo {
  map<string, string> attributes = 1;
}

message FooList {
  repeated Foo foos = 1;
}
```

and then in the `TestFooListStringer`, (and never in the `TestFooStringer` itself).

That only makes sense to me if the go runtime somehow sorts the map keys when printing it directly or in a struct, but not when the map is nested in a struct inside a list.

**Comment by romeovs** [comment]
To answer, we're on `0.4.0`.

**Comment by romeovs** [comment]
To elaborate, I put some debug statements in the `String` methods of both `Foo` and `FooList`. It seems that, although `FooList`s `String` calls:

```
fmt.Sprintf("%s", this.Foos)
```

the `Foo.String` method is not used there, which would explain the unordered map.

**Comment by romeovs** [comment]
To elaborate, when putting debug statements in the `String` methods of both `Foo` and `FooList`, it becomes clear that `FooList.String` never calls `Foo.String`, although that's what you would expect.

After all `FooList.String` calls:

```
fmt.Sprintf("%s", this.Foos)
```

weird stuff.

**Comment by awalterschulze** [comment]
Hmmm are you sure its not `%v` instead of `%s`, because I can't find %s in the stringer plugin?

**Comment by romeovs** [comment]
@awalterschulze Yes that's what I meant, sorry (updated the comment). The same issue stands however.

**Comment by awalterschulze** [comment]
%v is supposed to call String methods when it can find them :( and if it calls the String methods then the output should be deterministic.

**Comment by awalterschulze** [comment]
But I guess Go decided differently :(

**Comment by awalterschulze** [comment]
I cannot reproduce this failure with

```proto
syntax = "proto3";

package mapstringer;

import "github.com/gogo/protobuf/gogoproto/gogo.proto";

option (gogoproto.testgen_all) = true;
option (gogoproto.stringer_all) = true;
option (gogoproto.populate_all) = true;
option (gogoproto.equal_all) = true;
option (gogoproto.goproto_stringer_all) = false;

message RepMappy {
  repeated Mappy maps = 1;
}

message Mappy {
  map<string, string> kvs = 1;
}
```

**Comment by awalterschulze** [comment]
Makefile
```
regenerate:
	(protoc-min-version --version="3.0.0" --proto_path=../../../../../:../../protobuf/:. --gogofast_out=. mapstringer.proto)
```

**Comment by awalterschulze** [comment]
Could you try with the newest release version 0.5?

**Comment by romeovs** [comment]
Same issue in `v0.5`. I'll try to recreate the issue during the week!

**Comment by awalterschulze** [comment]
Thank you that would be great, because I couldn't recreate it.

**Comment by awalterschulze** [comment]
Have you had any success in recreating it?

**Comment by hubertp** [comment]
I think I'm hitting a variation of the reported issue.
https://github.com/gogo/protobuf/blob/master/proto/encode.go#L1157 means that whenever I want to serialize a message and the proto message contains a `map` field, it will not necessary return the same bytes. That was introduced a while ago in golang's protobuf (https://github.com/golang/protobuf/issues/86).
Would be cool if there was an option to configure that in the same way it is configurable in the java client (https://github.com/google/protobuf/blob/master/java/core/src/main/java/com/google/protobuf/CodedOutputStream.java#L158).

**Comment by awalterschulze** [comment]
If you generate a marshaler, then you can use the gogoproto.stable_marshaler extension to generate a marshaler where the maps are sorted for encoding.

This is different from this issue, which is about a protobuf message not printing a String deterministically.  But @romeovs has not given us a reproducing example and I couldn't reproduce it either.

**Comment by awalterschulze** [comment]
Can I close this?
